### PR TITLE
First pass at an update to 1.21.2.

### PR DESCRIPTION
**Pull request by gniftygnome**
This PR makes the API generally work from my testing so far.  The boat API is completely replaced under the hood, but the API interfaces are similar to before.  The main change is treating all boat and raft types of a single wood type as a group.  API calls use the group's Identifier and the API becomes the authoritative source of other identifiers.  Maybe some people won't like this; I think it is worth the convenience and simplicity.  One reasonable question is whether TerraformBoatData should be part of the API.

- Make it build against 1.21.2
- Fix broken things
- Replace basically all of the boat API

### CODE REVIEW

**Comment by gniftygnome** [comment]
Things I've noticed but don't yet understand:

* Now that they have a texture, I see the custom rafts do not render quite right.  There is a tiny one texture pixel high space above any edge viewed across the raft where you can see into the water.  I have absolutely no idea what is going on there.
* In just `MixinEatGrassGoal`, IDEA claims `org.spongepowered.asm.mixin.injection.callback.LocalCapture` does not exist (other classes in the same package do exist).  I have tried clearing caches and restarting and so on, but it persists.  However, this seems cosmetic; the build is fine and works fine...

**Comment by gniftygnome** [comment]
BTW I have Traverse fully updated to use this Terraform version, so if anybody wants that for testing, let me know and I'll push it.

File: gradle/wrapper/gradle-wrapper.properties
Comment (haykam821) [inline gradle/wrapper/gradle-wrapper.properties:1]: The Gradle version is updated here; should the other Gradle files be updated as well?

File: terraform-wood-api-v1/src/main/java/com/terraformersmc/terraform/boat/api/item/TerraformBoatItemHelper.java [hunk 1]
```diff
@@ -1,91 +1,101 @@
...
-	 * <p>To register a boat item and its dispenser behavior:
-	 * 
+	 * Creates and registers a {@linkplain BoatItem boat item} and associated
+	 * {@linkplain net.minecraft.entity.vehicle.BoatEntity boat entity} of the requested type
+	 * and with the default {@linkplain Item.Settings}.  This method assumes the type is a boat
+	 * instead of a raft.
+	 *
+	 * This method should be called once for each boat type.
+	 * Created items and entities will have identifiers similar to
+	 * {@code id.withSuffixedPath("_boat")} and {@code id.withPrefixedPath("chest_raft/")}.
```
Comment (haykam821) [inline terraform-wood-api-v1/src/main/java/com/terraformersmc/terraform/boat/api/item/TerraformBoatItemHelper.java:25]: Would the item ID be `examplemod:mahogany_boat` and the entity type ID be `examplemod:boat/mahogany`, then? I assume the second identifier refers to the entity model layer.

File: terraform-wood-api-v1/src/main/java/com/terraformersmc/terraform/wood/api/block/PillarLogHelper.java [hunk 2]
```diff
@@ -12,70 +12,66 @@ private PillarLogHelper() {
...
 	 * different map colors on the top/bottom versus the sides.
 	 *
 	 * @param wood Map color for non-bark faces of stem (ends)
 	 * @param bark Map color for bark faces of stem (sides)
-	 * @return New PillarBlock
+	 * @return New AbstractBlock.Settings
 	 */
-	public static PillarBlock ofNether(MapColor wood, MapColor bark) {
-		return new PillarBlock(AbstractBlock.Settings.create()
+	public static AbstractBlock.Settings settingsNether(MapColor wood, MapColor bark) {
```
Comment (haykam821) [inline terraform-wood-api-v1/src/main/java/com/terraformersmc/terraform/wood/api/block/PillarLogHelper.java]:
```suggestion
	public static AbstractBlock.Settings createNetherSettings(MapColor wood, MapColor bark) {
```

File: terraform-wood-api-v1/src/main/java/com/terraformersmc/terraform/wood/api/block/PillarLogHelper.java [hunk 3]
```diff
@@ -12,70 +12,66 @@ private PillarLogHelper() {
...
+	 * Factory to create default block settings for a PillarBlock Nether stem with
 	 * the same map color on all block faces.
 	 *
 	 * @param color Map color for all faces of stem
-	 * @return New PillarBlock
+	 * @return New AbstractBlock.Settings
 	 */
-	public static PillarBlock ofNether(MapColor color) {
-		return new PillarBlock(AbstractBlock.Settings.create()
+	public static AbstractBlock.Settings settingsNether(MapColor color) {
```
Comment (haykam821) [inline terraform-wood-api-v1/src/main/java/com/terraformersmc/terraform/wood/api/block/PillarLogHelper.java]:
```suggestion
	public static AbstractBlock.Settings createNetherSettings(MapColor color) {
```

File: terraform-wood-api-v1/src/main/java/com/terraformersmc/terraform/wood/api/block/PillarLogHelper.java [hunk 4]
```diff
@@ -12,70 +12,66 @@ private PillarLogHelper() {
...
 	 * different map colors on the top/bottom versus the sides.
 	 *
 	 * @param wood Map color for non-bark faces of log (ends)
 	 * @param bark Map color for bark faces of log (sides)
-	 * @return New PillarBlock
+	 * @return New AbstractBlock.Settings
 	 */
-	public static PillarBlock of(MapColor wood, MapColor bark) {
-		return new PillarBlock(AbstractBlock.Settings.create()
+	public static AbstractBlock.Settings settings(MapColor wood, MapColor bark) {
```
Comment (haykam821) [inline terraform-wood-api-v1/src/main/java/com/terraformersmc/terraform/wood/api/block/PillarLogHelper.java]:
```suggestion
	public static AbstractBlock.Settings createSettings(MapColor wood, MapColor bark) {
```

File: terraform-wood-api-v1/src/main/java/com/terraformersmc/terraform/wood/api/block/PillarLogHelper.java [hunk 5]
```diff
@@ -12,70 +12,66 @@ private PillarLogHelper() {
...
+	 * Factory to create default block settings for a PillarBlock log with
 	 * the same map color on all block faces.
 	 *
 	 * @param color Map color for all faces of log
-	 * @return New PillarBlock
+	 * @return New AbstractBlock.Settings
 	 */
-	public static PillarBlock of(MapColor color) {
-		return new PillarBlock(AbstractBlock.Settings.create()
+	public static AbstractBlock.Settings settings(MapColor color) {
```
Comment (haykam821) [inline terraform-wood-api-v1/src/main/java/com/terraformersmc/terraform/wood/api/block/PillarLogHelper.java]:
```suggestion
	public static AbstractBlock.Settings createSettings(MapColor color) {
```

File: gradle/wrapper/gradle-wrapper.properties
Comment (gniftygnome) [reply gradle/wrapper/gradle-wrapper.properties:1]: I don't usually bother, but sure, I can go refresh myself on how to do that.

File: terraform-wood-api-v1/src/main/java/com/terraformersmc/terraform/boat/api/item/TerraformBoatItemHelper.java (hunk 1 above)
Comment (gniftygnome) [reply terraform-wood-api-v1/src/main/java/com/terraformersmc/terraform/boat/api/item/TerraformBoatItemHelper.java:25]: Yes.  Now that we are doing this entirely the vanilla way, we create an item, entity type, and model layer.  For our mods, there is no need to know the identifiers (although they are predictable, as needed to put the assets in the right place).  I doubt anybody needs the model layer.  I have wondered if others might need either the identifiers or the entity type.  If we think folks might need things like these, I could refactor so the getters of TerraformBoatData are exposed in the API.

File: gradle/wrapper/gradle-wrapper.properties
Comment (gniftygnome) [reply gradle/wrapper/gradle-wrapper.properties:1]: Done in  [1448a67](https://github.com/TerraformersMC/Terraform/pull/95/commits/1448a6769f1629dc745202fe9d4bb233b9e90d12).

**Comment by gniftygnome** [comment]
> * Now that they have a texture, I see the custom rafts do not render quite right.  There is a tiny one texture pixel high space above any edge viewed across the raft where you can see into the water.  I have absolutely no idea what is going on there.

This is fixed now in [1248758](https://github.com/TerraformersMC/Terraform/pull/95/commits/1248758da4677b49e90d207d4b57aec22089d830).

**Comment by gniftygnome** [comment]
OK, I think the API is fine; perhaps things could have better names but we can always fix that in 1.21.4 or whatever.  If we do want to improve this in 1.21.2/3, it would be via API additions or internal changes, which shouldn't be disruptive.  I'll release this as `12.0.0-alpha.1`.

**Pull request merged**